File: core/src/pl/shockah/godwit/gl/color/RGBColorSpace.java
package pl.shockah.godwit.gl.color;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.math.MathUtils;

import javax.annotation.Nonnull;

import lombok.EqualsAndHashCode;
import pl.shockah.godwit.fx.ease.Easing;

@EqualsAndHashCode
public class RGBColorSpace implements ColorSpace<RGBColorSpace> {
	public float r;
	public float g;
	public float b;

	public RGBColorSpace(float r, float g, float b) {
		this.r = r;
		this.g = g;
		this.b = b;
	}

	@Nonnull public static RGBColorSpace from(@Nonnull Color color) {
		return new RGBColorSpace(color.r, color.g, color.b);
	}

	@Override
	public String toString() {
		return String.format("[RGBColorSpace: R:%.3f G:%.3f B:%.3f]", r, g, b);
	}

	@Override
	@Nonnull public RGBColorSpace toRGB() {
		return this;
	}

	@Override
	public float getDistance(@Nonnull RGBColorSpace other) {
		return (float)Math.sqrt(
				Math.pow(r - other.r, 2)
						+ Math.pow(g - other.g, 2)
						+ Math.pow(b - other.b, 2)
		);
	}

	@Override
	@Nonnull public Color toColor() {
		return new Color(
				MathUtils.clamp(r, 0f, 1f),
				MathUtils.clamp(g, 0f, 1f),
				MathUtils.clamp(b, 0f, 1f),
				1f
		);
	}

	@Override
	@Nonnull public RGBColorSpace ease(@Nonnull RGBColorSpace other, float f) {
		return new RGBColorSpace(
				Easing.linear.ease(r, other.r, f),
				Easing.linear.ease(g, other.g, f),
				Easing.linear.ease(b, other.b, f)
		);
	}
}

File: core/src/pl/shockah/godwit/color/RGBColorSpace.kt
package pl.shockah.godwit.color

import kotlin.math.pow
import kotlin.math.sqrt

data class RGBColorSpace(
		var r: Float,
		var g: Float,
		var b: Float
) : ColorSpace<RGBColorSpace> {
	override val rgb: RGBColorSpace
		get() = copy()

	override fun copy(): RGBColorSpace = RGBColorSpace(r, g, b)

	override fun getDistance(other: RGBColorSpace): Float {
		return sqrt((r - other.r).pow(2) + (g - other.g).pow(2) + (b - other.b).pow(2))
	}

	override fun ease(other: RGBColorSpace, f: Float): RGBColorSpace {
		return RGBColorSpace(
				r.ease(other.r, f),
				g.ease(other.g, f),
				b.ease(other.b, f)
		)
	}
}